### PR TITLE
Check cURL response codes and do not deliver data on error

### PR DESCRIPTION
**Pull request by IyeOnline**
This solves issues with the http loader successfully yielding a 404 page without an error report.

Fixes https://github.com/tenzir/issues/issues/2202

### CODE REVIEW

File: libtenzir/src/curl.cpp [hunk 1]
```diff
@@ -226,6 +226,17 @@ auto multi::run(std::chrono::milliseconds timeout) -> caf::expected<size_t> {
   if (result != code::ok) {
     return to_error(result);
   }
+  auto queued = int{};
+  auto* msg = curl_multi_info_read(multi_.get(), &queued);
+  if (msg) {
+    auto response_code = int64_t{0};
+    curl_easy_getinfo(msg->easy_handle, CURLINFO_RESPONSE_CODE, &response_code);
```
Comment (mavam) [inline libtenzir/src/curl.cpp:233]: I think you need to check the return code and perform the subsequent response code check only after you got `code::ok`.

**Review by mavam — CHANGES_REQUESTED** [review]
The check might do what it's supposed to, but it's at the wrong location. Since cURL is a multi-protocol tool, this check must only be performed in the case where we actually got a (HTTP) response code.

Let's discuss where this would be better.

**Comment by mavam** [comment]
@IyeOnline I orthogonalized the curl and transfer abstractions again, keeping the former just as an unopinionated wrapper and the latter the thing where we opine.

(I didn't get a chance to test this yet, though, as I'm still traveling. Mind giving it a try?)

**Review by mavam — APPROVED** [review]
The response logic itself is fine. I moved it to a different location, though.

(Self-approving modulo testing.)

File: libtenzir/src/curl.cpp [hunk 2]
```diff
@@ -60,6 +60,31 @@ easy::easy() : easy_{curl_easy_init()} {
   TENZIR_ASSERT(easy_ != nullptr);
 }
 
+auto easy::get(info i) -> std::pair<code, info_data> {
+  // TODO: port missing info enums as you see fit. Do this, for example, with a
+  // `man CURLINFO_X` and check the result type and then adding the missing
+  // case.
```
Comment (dominiklohmann) [inline libtenzir/src/curl.cpp:66]: This looks like it still needs to be done.

File: libtenzir/src/curl.cpp [hunk 3]
```diff
@@ -60,6 +60,31 @@ easy::easy() : easy_{curl_easy_init()} {
...
+      auto c = curl_easy_getinfo(easy_.get(), curl_info, &url);
+      return {static_cast<code>(c), std::string_view{url}};
+    }
+    case info::response_code: {
+      long* response_code = nullptr;
+      auto c = curl_easy_getinfo(easy_.get(), curl_info, &response_code);
+      return {static_cast<code>(c), *response_code};
+    }
+  }
+  __builtin_unreachable();
```
Comment (dominiklohmann) [inline libtenzir/src/curl.cpp:85]:
```suggestion
  TENZIR_UNREACHABLE();
```

File: libtenzir/src/transfer.cpp [hunk 4]
```diff
@@ -220,6 +220,21 @@ auto transfer::download_chunks() -> generator<caf::expected<chunk_ptr>> {
...
+      auto [code, info] = easy_.get(curl::easy::info::response_code);
+      if (code == curl::easy::code::ok) {
+        TENZIR_ASSERT(std::holds_alternative<long>(info));
+        // FTP, HTTP and SMTP, error codes in [200,299] are okay. Augment this
+        // check to support more protocols.
+        auto response_code = std::get<long>(info);
+        if (response_code < 200 or response_code > 299) {
+          co_yield caf::make_error(ec::invalid_result,
+                                   fmt::format("HTTP response code: {}",
+                                               response_code));
```
Comment (dominiklohmann) [inline libtenzir/src/transfer.cpp:233]: This should use `diagnostic::error(…).to_error()` like all new code.

File: libtenzir/include/tenzir/curl.hpp [hunk 5]
```diff
@@ -181,8 +181,93 @@ class easy {
...
+    capath = CURLINFO_CAPATH,
+    xfer_id = CURLINFO_XFER_ID,
+    conn_id = CURLINFO_CONN_ID,
+    queue_time_t = CURLINFO_QUEUE_TIME_T,
+    used_proxy = CURLINFO_USED_PROXY,
+    lastone = CURLINFO_LASTONE
+  };
+
+  /// Possible return type data when calling `get_info`.
+  using info_data = std::variant<bool, long, std::string_view>;
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/curl.hpp:264]: The comment should really explain what it means to get a bool, long, or string view, respectively.

**Pull request merged**